testcases/kernel/syscalls/fanotify/fanotify09.c

```diff
@@ -0,0 +1,236 @@
+		 * The one mark with FAN_EVENT_ON_CHILD is needed for
+		 * setting the DCACHE_FSNOTIFY_PARENT_WATCHED dentry
+		 * flag.
+		 */
+		onchild = (i == 0) ? FAN_EVENT_ON_CHILD : 0;
+		ret = fanotify_mark(fd_notify[i],
+				    FAN_MARK_ADD,
+				    FAN_MODIFY | onchild, AT_FDCWD, ".");
+		if (ret < 0) {
+			tst_brk(TBROK | TERRNO,
+				 "fanotify_mark(%d, FAN_MARK_ADD, "
+				 "FAN_MODIFY%s, AT_FDCWD, '.') failed",
+				 fd_notify[i],
+				 onchild ? " | FAN_EVENT_ON_CHILD" : "");
+		}
+	}
+}
+
+static void cleanup_fanotify_groups(void)
+{
+	unsigned int i;
+
+	for (i = 0; i < NUM_GROUPS; i++) {
+		if (fd_notify[i] && fd_notify[i] != -1) {
+			if (close(fd_notify[i]) == -1)
+				tst_res(TWARN, "close(%d) failed",
+					 fd_notify[i]);
+			fd_notify[i] = 0;
+		}
+	}
+}
+
+static void verify_event(int group, struct fanotify_event_metadata *event)
+{
+	if (event->mask != FAN_MODIFY) {
+		tst_res(TFAIL, "group %d get event: mask %llx (expected %llx) "
+			 "pid=%u fd=%u", group, (unsigned long long)event->mask,
+			 (unsigned long long)FAN_MODIFY,
+			 (unsigned)event->pid, event->fd);
+	} else if (event->pid != getpid()) {
+		tst_res(TFAIL, "group %d get event: mask %llx pid=%u "
+			 "(expected %u) fd=%u", group,
+			 (unsigned long long)event->mask, (unsigned)event->pid,
+			 (unsigned)getpid(), event->fd);
+	} else {
+		tst_res(TPASS, "group %d get event: mask %llx pid=%u fd=%u",
+			 group, (unsigned long long)event->mask,
+			 (unsigned)event->pid, event->fd);
+	}
+}
+
+void test01(void)
+{
+	int ret;
+	unsigned int i;
+	struct fanotify_event_metadata *event;
+
+	create_fanotify_groups();
+
+	/*
+	 * generate MODIFY event and no FAN_CLOSE_NOWRITE event.
+	 */
+	SAFE_FILE_PRINTF(fname, "1");
+
+	/* First verify the first group got the MODIFY event */
+	ret = read(fd_notify[0], event_buf, EVENT_BUF_LEN);
+	if (ret < 0) {
+		if (errno == EAGAIN) {
+			tst_res(TFAIL, "first group did not get event");
+		} else {
+			tst_brk(TBROK | TERRNO,
+				 "reading fanotify events failed");
+		}
+	}
+	if (ret < (int)FAN_EVENT_METADATA_LEN) {
+		tst_brk(TBROK,
+			 "short read when reading fanotify "
+			 "events (%d < %d)", ret,
+			 (int)EVENT_BUF_LEN);
+	}
+	event = (struct fanotify_event_metadata *)event_buf;
+	if (ret > (int)event->event_len) {
+		tst_res(TFAIL, "first group got more than one "
+			 "event (%d > %d)", ret,
+			 event->event_len);
+	} else {
+		verify_event(0, event);
+	}
+	if (event->fd != FAN_NOFD)
+		close(event->fd);
+
+	/* Then verify the rest of the groups did not get the MODIFY event */
+	for (i = 1; i < NUM_GROUPS; i++) {
+		ret = read(fd_notify[i], event_buf, EVENT_BUF_LEN);
+		if (ret > 0) {
+			tst_res(TFAIL, "group %d got event", i);
+			if (event->fd != FAN_NOFD)
+				close(event->fd);
+		} else if (ret == 0) {
+			tst_brk(TBROK, "zero length "
+				 "read from fanotify fd");
+		} else if (errno != EAGAIN) {
+			tst_brk(TBROK | TERRNO,
+				 "reading fanotify events failed");
+		} else {
```

Comment:
Just a note, the `tst_brk()` will exit the test execution immediately (it calls cleanup() then the library does it's cleanup then it calls exit()), hence there is no need for any else branches.